crates/matrix-sdk-ui/src/room_list/mod.rs

```diff
@@ -0,0 +1,620 @@
+//! specific usage of Sliding Sync.
+//!
+//! # Basic principle
+//!
+//! `RoomList` works with 2 Sliding Sync List:
+//!
+//! * `all_rooms` (refered by the constant [`ALL_ROOMS_LIST_NAME`]) is the main
+//!   list. It's goal is to load all the user' rooms. It starts with a
+//!   [`SlidingSyncMode::Selective`] sync-mode with a small range (i.e. a small
+//!   set of rooms) to load the first rooms quickly, and then updates to a
+//!   [`SlidingSyncMode::Growing`] sync-mode to load the remaining rooms “in the
+//!   background”: it will sync the existing rooms and will fetch new rooms, by
+//!   a certain batch size.
+//! * `visible_rooms` (refered by the constant [`VISIBLE_ROOMS_LIST_NAME`]) is
+//!   the “reactive” list. It's goal is to react to the client app user actions.
+//!   If the user scrolls in the room list, the `visible_rooms` will be
+//!   configured to sync for the particular range of rooms the user is actually
+//!   seeing (the rooms in the current viewport). `visible_rooms` has a
+//!   different configuration than `all_rooms` as it loads more timeline events:
+//!   it means that the room will already have a “history”, a timeline, ready to
+//!   be presented when the user enters the room.
+//!
+//! This behavior has proven to be empirically satisfying to provide a fast and
+//! fluid user experience for a Matrix client.
+//!
+//! [`RoomList::entries`] provides a way to get a stream of room list entry.
+//! This stream can be filtered, and the filter can be changed over time.
+//!
+//! [`RoomList::state_stream`] provides a way to get a stream of the state
+//! machine's state, which can be pretty helpful for the client app.
+
+use std::{future::ready, sync::Arc};
+
+use async_stream::stream;
+use async_trait::async_trait;
+use eyeball::shared::Observable;
+use eyeball_im::VectorDiff;
+use futures_util::{pin_mut, Stream, StreamExt};
+use imbl::Vector;
+pub use matrix_sdk::RoomListEntry;
+use matrix_sdk::{
+    sliding_sync::Ranges, Client, Error as SlidingSyncError, SlidingSync, SlidingSyncList,
+    SlidingSyncMode, SlidingSyncRoom,
+};
+use once_cell::sync::Lazy;
+use ruma::{OwnedRoomId, RoomId};
+use thiserror::Error;
+
+use crate::{timeline::EventTimelineItem, Timeline};
+
+pub const ALL_ROOMS_LIST_NAME: &str = "all_rooms";
+pub const VISIBLE_ROOMS_LIST_NAME: &str = "visible_rooms";
+
+/// The [`RoomList`] type. See the module's documentation to learn more.
+#[derive(Debug)]
+pub struct RoomList {
+    sliding_sync: SlidingSync,
+    state: Observable<State>,
+}
+
+impl RoomList {
+    /// Create a new `RoomList`.
+    ///
+    /// A [`matrix_sdk::SlidingSync`] client will be created, with a cached list
+    /// already pre-configured.
+    pub async fn new(client: Client) -> Result<Self, Error> {
+        let sliding_sync = client
+            .sliding_sync()
+            .storage_key(Some("matrix-sdk-ui-roomlist".to_string()))
+            .add_cached_list(
```

Comment:
:warning: This is what caused the slowness on startup of ElementX iOS, since the `ALL_ROOMS` list is likely to contain many lists and rooms, and reloading from cache + building a timeline will attempt to decrypt all the previous messages present in those rooms.

Perhaps disable with a big `TODO` that it should be added back in the future, when the timeline is properly cached?

Reply:
Let's address in another PR. The goal of this PR is not to fix this problem.

Reply:
Just to make sure we're on the same page: I am fine with that, but that means we can't use the new Room API until we have 1. either made restoring much faster, 2. or replaced this line with `add_list` until we do (1). I'm suggesting we could do (2) right now, so as to be able to use the Room API as soon as possible in the apps.